.clang-format

```diff
@@ -0,0 +1,24 @@
+BasedOnStyle: LLVM
+BraceWrapping:
+  SplitEmptyFunction: false
+  SplitEmptyRecord: false
+IndentWidth: 3
+ColumnLimit: 120
+PointerAlignment: Left
+AccessModifierOffset: -3
+AlwaysBreakTemplateDeclarations: true
+AlignArrayOfStructures: Left
```

Comment:
For context, here are the default settings in the predefined styles:
```
AlwaysBreakTemplateDeclarations: Yes  (chromium) 
AlwaysBreakTemplateDeclarations: MultiLine  (gnu) 
AlwaysBreakTemplateDeclarations: Yes  (google) 
AlwaysBreakTemplateDeclarations: Yes  (leap_proposed) 
AlwaysBreakTemplateDeclarations: MultiLine  (llvm) 
AlwaysBreakTemplateDeclarations: MultiLine  (microsoft) 
AlwaysBreakTemplateDeclarations: Yes  (mozilla) 
AlwaysBreakTemplateDeclarations: MultiLine  (webkit) 
```

Reply:
@kj4ezj: Thanks for pointing out that it should be Yes instead of true.

As for switching it to No, I'm pretty strongly against that one. So unless there are a lot of other people who think it should be Yes, I say we simply correct the setting by changing it from `true` to `Yes`.

Reply:
> @kj4ezj: Thanks for pointing out that it should be Yes instead of true.

No, thank you!

> As for switching it to No, I'm pretty strongly against that one.

Can you please share your justification for restricting syntax here in the context of our stated goal of improving readability?

You stated your preference and I deeply respect your domain experience but, without any context as to why that is your preference, I can't learn or change my opinion. My default is to oppose restrictions on syntax and expression, and to leave these decisions to developers. I would like to understand your position.

Reply:
@kj4ezj: It all comes down to matters of taste. There isn't an objectively correct answer here. But I think the below examples show why I consider the formatting to be less consistent with `AlwaysBreakTemplateDeclarations: No` compared to the other options.

With `AlwaysBreakTemplateDeclarations: No`:

```cpp
template <typename T> struct some_type_name {};

template <typename T, typename U>
void swap_function_with_a_really_long_name(some_type_name<T>& lhs, some_type_name<U>& rhs) {
   // TODO: implement
}

template <typename T, typename U> void swap_function_with_shorter_name(some_type_name<T>& lhs, some_type_name<U>& rhs) {
   // TODO: implement
}

template <typename TemplateParameter1,
          typename TemplateParameter2,
          typename TemplateParameter3,
          typename TemplateParameter4>
int some_other_function(int x) {
   // TODO: implement
   return 0;
}

namespace namespace_name_adds_additional_length {

struct on_top_of_already_long_return_type_name {};

} // namespace namespace_name_adds_additional_length

template <typename T, typename U> namespace_name_adds_additional_length::on_top_of_already_long_return_type_name
another_swap_function(some_type_name<T>& lhs, some_type_name<U>& rhs) {
   // TODO: implement
   return {};
}

template <typename T, typename U> auto yet_another_swap_function(some_type_name<T>& lhs, some_type_name<U>& rhs)
   -> namespace_name_adds_additional_length::on_top_of_already_long_return_type_name {
   // TODO: implement
   return {};
}
```

Note the difficulty in quickly finding the return type of the swap function depending on how long the name of the function is (plus the length of other arguments relative to the column limit). 

Usually, the template parameters take up a small enough amount of space to fit on one line. So if `AlwaysBreakTemplateDeclarations` was not `No`, then the eye can quickly scan for the template line, skip that, and then the next line should usually start with the return type of the function unless `auto` is used. If auto is used, then the reader would have to wait until scanning through the parameters of the function before seeing what the return value is following the `->`.

But with `AlwaysBreakTemplateDeclarations: No`, you can see how the return type of the `swap_function_with_shorter_name` function is sandwiched in the middle and harder to pick out. Similar problem exists even with identifying the `auto` in the case of the `yet_another_swap_function` function.

For the `another_swap_function` function, the return type is not sandwiched between the template parameters and the function name, but the long return type is still not starting as the first thing on its own line which still makes it harder to quickly identify.

With `AlwaysBreakTemplateDeclarations: MultiLine` (this is the one in the LLVM preset):

```cpp
template <typename T> struct some_type_name {};

template <typename T, typename U>
void swap_function_with_a_really_long_name(some_type_name<T>& lhs, some_type_name<U>& rhs) {
   // TODO: implement
}

template <typename T, typename U> void swap_function_with_shorter_name(some_type_name<T>& lhs, some_type_name<U>& rhs) {
   // TODO: implement
}

template <typename TemplateParameter1,
          typename TemplateParameter2,
          typename TemplateParameter3,
          typename TemplateParameter4>
int some_other_function(int x) {
   // TODO: implement
   return 0;
}

namespace namespace_name_adds_additional_length {

struct on_top_of_already_long_return_type_name {};

} // namespace namespace_name_adds_additional_length

template <typename T, typename U>
namespace_name_adds_additional_length::on_top_of_already_long_return_type_name
another_swap_function(some_type_name<T>& lhs, some_type_name<U>& rhs) {
   // TODO: implement
   return {};
}

template <typename T, typename U>
auto yet_another_swap_function(some_type_name<T>& lhs, some_type_name<U>& rhs)
   -> namespace_name_adds_additional_length::on_top_of_already_long_return_type_name {
   // TODO: implement
   return {};
}
```

Now there is an improvement in readability (IMO of course) for the `another_swap_function` and `yet_another_swap_function` functions. 

But the same problem as before persists for the `swap_function_with_shorter_name` function.

Additionally, I don't love that a quick vertical scan makes it difficult to identify the `struct` template distinguished from other template entities like template functions.

Both of those issues are resolved by going with `AlwaysBreakTemplateDeclarations: Yes`.

With `AlwaysBreakTemplateDeclarations: Yes`:

```cpp
template <typename T>
struct some_type_name {};

template <typename T, typename U>
void swap_function_with_a_really_long_name(some_type_name<T>& lhs, some_type_name<U>& rhs) {
   // TODO: implement
}

template <typename T, typename U>
void swap_function_with_shorter_name(some_type_name<T>& lhs, some_type_name<U>& rhs) {
   // TODO: implement
}

template <typename TemplateParameter1,
          typename TemplateParameter2,
          typename TemplateParameter3,
          typename TemplateParameter4>
int some_other_function(int x) {
   // TODO: implement
   return 0;
}

namespace namespace_name_adds_additional_length {

struct on_top_of_already_long_return_type_name {};

} // namespace namespace_name_adds_additional_length

template <typename T, typename U>
namespace_name_adds_additional_length::on_top_of_already_long_return_type_name
another_swap_function(some_type_name<T>& lhs, some_type_name<U>& rhs) {
   // TODO: implement
   return {};
}

template <typename T, typename U>
auto yet_another_swap_function(some_type_name<T>& lhs, some_type_name<U>& rhs)
   -> namespace_name_adds_additional_length::on_top_of_already_long_return_type_name {
   // TODO: implement
   return {};
}
```

To me the above version seems more consistent compared to the previous two versions.


Incidentally, I think the above example also demonstrates why I think it is also nice to have `AfterFunction: true` in the custom `BraceWrapping`. In fact, I prefer the following custom `BraceWrapping`: https://github.com/AntelopeIO/reference-contracts/blob/efff1254806aa2df68d06f5d5e4b5c122297a857/.clang-format#L13-L27

But anyway, that additional point on `BraceWrapping` changes is unrelated to this conversation thread, and it is not something that I care about strongly enough to advocate for that change if other people are fine with the custom brace wrapping that already exists in this PR (i.e. the one that has more minimal changes compared to the LLVM preset).

Reply:
Yes, braces after function help with the situation I mentioned earlier too https://github.com/AntelopeIO/leap/pull/652#discussion_r1095163303

Reply:
I also much prefer `BraceWrapping`  `After*: true`, but I didn't originally advocate for it as it didn't seem to be used in a preponderance of the leap code I saw. I feel it makes it easier to see the quickly functions/structs/classes .

Reply:
I also prefer the `Align*` we removed. Yes, it may cause larger diffs, but there are ways around it:
1. when we have multiple declarations in a row, it is easy to pick a variable name that is not longer than the longest current one.
2. if we want a longer variable, you can just put the new declaration last and add an empty line before it, and it will not be aligned with the previous group. 

In any case, with clang-format, there will be cases where a change causes whitespace change on the lines around the change, I think it is unavoidable.

Reply:
The [.clang_format](https://github.com/AntelopeIO/reference-contracts/blob/efff1254806aa2df68d06f5d5e4b5c122297a857/.clang-format)  that Areg linked really looks good I think, and it seems that we are converging towards something close to it.

Reply:
Thank you @arhag for sharing the context on why you prefer `AlwaysBreakTemplateDeclarations: Yes`. I generally prefer not to impose more restrictions on developers and code expression. However, that is a good and thorough justification, and I do see how this could improve readability. I don't think we should target these decisions to specific tooling, but I can see how this rule would especially improve the readability of templates in the VScode "peek" functionality that I personally use.

I will mark this thread as "resolved" and leave this up to the rest of the team. If someone else feels strongly, please feel free to "un-resolve" this thread and continue the discussion.

Reply:
As a side-node, albeit unrelated to this thread, since you guys brought it up...

> Incidentally, I think the above example (specifically `yet_another_swap_function` with the `auto` return type) also demonstrates why I think it is also nice to have `AfterFunction: true` in the custom `BraceWrapping`.

> Yes, braces after function help with the situation I mentioned earlier too [#652 (comment)](https://github.com/AntelopeIO/leap/pull/652#discussion_r1095163303)

> I also much prefer `BraceWrapping` with `After*: true` (for non-control statements) as specified in Areg's link above, but I didn't originally advocate for it as it didn't seem to be used in a preponderance of the leap code I saw. I feel it makes it easier to see the quickly functions/structs/classes.

...I also prefer braces to be on their own line in C-like languages.
```c++
void foo()
{
  bar();
  bar2();
}
```
I have always felt it makes code more readable to have braces lined up in C-like languages, despite the fact that JavaScript has partly corrupted me in this sense. I would 100% support turning this on, not that it is up to me.

RE: @greg7mdp @arhag @spoonincode

Reply:
I updated the .clang-format to Wrap braces after [Class, Enum, Function, Struct, Union]. It is now quite close to the [.clang_format](https://github.com/AntelopeIO/reference-contracts/blob/efff1254806aa2df68d06f5d5e4b5c122297a857/.clang-format)  from reference-contracts.

Reply:
I dislike all the Wrap braces after. To me it is an unnecessary waste of a line and it breaks up the flow. Particularly for functions.  Visibility distances the implementation of the function from the function declaration. This is my least-favorite deviation from current practice.

Do we really want:
```c++
std::string foo() const
{
   return s;
}
```

Over:
```c++
std::string foo() const {
   return s;
}
```
Or even:
```c++
std::string foo() const { return s; }
```

Reply:
Yes Kevin, I also feel this is the biggest issue with this setting is with short functions.  But then I like `Wrap braces after` for struct and classes, and I think it is nice for longer functions, so I think the extra line is worth it for the occasional short function which doesn't fit on one line.

For your last example (very short function), we can set `AllowShortFunctionsOnASingleLine: All` , which I also have in my personal .clang-format. We currently have `Inline` for that setting, so it would apply only for functions defined inside classes.

I actually have the following in my personal .clang-format:

```
AllowShortBlocksOnASingleLine: Empty
AllowShortCaseLabelsOnASingleLine: true
AllowShortFunctionsOnASingleLine: All
AllowShortLambdasOnASingleLine: All
AllowShortIfStatementsOnASingleLine: Never
AllowShortLoopsOnASingleLine: false
```

Reply:
I did change `AllowShortFunctionsOnASingleLine` from `Inline` to `All`